### PR TITLE
(#4372) Pruning rev_map at compaction for leveldb

### PR DESCRIPTION
**Pull request by sgenoud**
We only keep the revision - sequence map for revisions that have a
document in the database.

### CODE REVIEW

**Comment by sgenoud** [comment]
My first PR - I did not know what to add as a test, feedback welcome!

**Comment by nolanlawson** [comment]
I'm afraid this doesn't really fix #4372 - this only changes the behavior when `get`ting documents from the database, while leaving unresolved the problem that too much data stays written to the database.

The real fix would be to remove the old revs when they are inserted into the database, although unfortunately it's tricky to write a test for that, because the information isn't really exposed by the API in any way (the only way you'd know is by looking at the disk space used by the database). So ultimately if/when we do get a fix, I think a fix without a test would be fine.

The other tricky part of this fix would be fixing it for both the case of compaction (which is a manual operation done by the `compactDocument` function) and auto-compaction (an operation done as part of `bulkDocs`).

**Comment by nolanlawson** [comment]
Also it's worth noting that the `rev_map` is merely a map of rev strings to integer seqs, meaning it doesn't actually contain the document data (although ideally it would be pruned as well).

**Comment by sgenoud** [comment]
> I'm afraid this doesn't really fix #4372 - this only changes the behavior when getting documents from the database

I am not sure that I understands what you mean. Here is how I understand #4372 : it is two fold:

1. When a document reaches 1000 revisions it starts forgetting about the old ones (stemming) - but only in the rev_tree - but the actual documents should also be removed from the database. This is actually a problem not limited to the leveldb adapter. 

2. If autocompact is on, the first issue should not appear, but as reported the leveldb file size keeps getting bigger anyway. The reason for that is that on `doCompact` the `rev_map` is not updated - i.e. the document entry is removed, but the map keeps getting bigger (in a normal case with not that many revisions it is not an issue, but for the particular case of many many revisions it grows in size).

Anyway, I have implemented a solution to problem `1` (in the three different adapters) as well. I have seen that it fixes the leveldb issue but did not try to reproduce on the other adapters.

**Comment by nolanlawson** [comment]
OK, nice writeup, and I see that you have some updates now that address the issue of other adapters. I'll take a look, and restart any failed tests (Travis typically flakes out on a couple tests; we can just restart them).

File: src/adapters/leveldb/index.js [hunk 1]
```diff
@@ -571,11 +570,20 @@ function LevelPouch(opts, callback) {
...
+      fetchedDocs.forEach(function (metadata, docId) {
+        revsMap.set(docId, compactTree(metadata));
+      });
+      compact(revsMap, callback);
+    }
+
     function finish() {
       if (api.auto_compaction) {
         return autoCompact(complete);
+      } else {
```
Comment (daleharvey) [inline src/adapters/leveldb/index.js]: Small nit, everywhere else in the code has 2 space indents and we would like to be consistent

**Comment by daleharvey** [comment]
This looks great, its a pretty important issues so thanks a lot for taking a look at it. Sorry for taking so long to get round to review (holidays + moving home). 

Theres a few minor nits around the indentation but the code in general looks good. What we really need now is a way to test this which is going to be pretty tricky to do consistently, theres going to be 2 main ways, is there any way these 'leaks' are going to be exposed at the api level (open_revs etc) if so then we can do that. Otherwise we will likely be stuck with node only tests that check the database size (or even contents).

I will take another look at how to test this in the next few days, thats the only thing thats blocking this from landing as far as I can see, Cheers

**Comment by sgenoud** [comment]
>  Sorry for taking so long to get round to review (holidays + moving home).

No problem!

> is there any way these 'leaks' are going to be exposed at the api level (open_revs etc) if so then we can do that in integration tests.

I added a test, based on @nolanlawson ideas in the original issue: revisions that are stemmed should not be available to directly get - I could reproduce the error with the test on master, while it passes on the fix branch.

It allowed me to find a bug for the websql implementation (I guess it is what was caught by travis), yeah!

> Theres a few minor nits around the indentation but the code in general looks good. 

Additionally, the version of jshint that we use do not catch many of the coding style errors - I brought that up on the Slack, should I open a separate issue for this?

**Comment by daleharvey** [comment]
> Additionally, the version of jshint that we use do not catch many of the coding style 
> errors - I brought that up on the Slack, should I open a separate issue for this?

Definitely, the better job our linter can do the easier it is for us / constributors.

Will take a proper look at this tomorrow, but initially its looking great

**Comment by daleharvey** [comment]
ok took another look, sorry I missed the test first time that is great, I still think we could do with another test that confirms no storage leaks but I think this is a great start. We cant copy couch's way of stemming those revs so I think this is the best way. :+1: thanks for the PR :)

**Comment by daleharvey** [comment]
Merged in https://github.com/pouchdb/pouchdb/commit/4cccf617bce67351123a842f12ff0e694587f111

**Pull request closed without merging**